Ref count various WebKit types within Swift

adetaylor opened the pull request:
#### 85b8b03d430589e9a39c93e69983ef393e17fc8a
<pre>
Ref count various WebKit types within Swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=302644">https://bugs.webkit.org/show_bug.cgi?id=302644</a>
<a href="https://rdar.apple.com/164888628">rdar://164888628</a>

Reviewed by Geoffrey Garen.

This informs Swift that some C++ types should be treated as Swift reference
types (i.e. reference-counted) rather than value types. It teaches Swift
how to manipulate the reference counts for these types.

This is a temporary solution. In due course, we&apos;ll be able to inform
Swift that the base WTF::RefCounted type is the ancestor of all reference
counted types, and it won&apos;t be necessary to make these changes to
individual classes.

That&apos;s enabled by either of these changes:
<a href="https://rdar.apple.com/131065240">rdar://131065240</a> - allow templated types to be ref counted
<a href="https://rdar.apple.com/160696723">rdar://160696723</a> - allow method calls for ref and deref
We can&apos;t currently use this functionality in WebKit because we need
to support older compilers, and also due to a compiler
bug <a href="https://rdar.apple.com/162195228">rdar://162195228</a>. So for now, we will annotate just those
specific types which are required for upcoming CoreIPC message handler
work.

To avoid warnings/errors, these functions need to have their parameters
marked as _Nonnull, or specifically to adopt the WTF macro WTF_NONNULL.
That currently causes problems with -Wnullability-completeness, which
complains that other functions in the same headers are not duly
annotated. WebKit has switched off this warning for its whole codebase
but had overlooked TestWebKitAPI, which does not inherit warning flags
from the rest of the .xcconfig system. Add this warning flag there too
in order to prevent these compile errors when TestWebKitAPI indirectly
includes IPC/Connection.h

Canonical link: <a href="https://commits.webkit.org/303461@main">https://commits.webkit.org/303461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49337756612d9141c84e9d3d96eabccb3cf669a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139583 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83980 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f0c1d04b-cbd9-4cb3-aef2-9143293baf93) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133939 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4323 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100957 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68315 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4363c9c5-6ff9-4938-ac6d-03127f0b1c38) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3207 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118291 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81748 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a0661a02-fbed-4499-be7d-9249c767694e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3104 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1100 "Found 2 new API test failures: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer, TestWebKitAPI.WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82803 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111874 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142230 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4231 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109327 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4312 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109500 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27834 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3211 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114568 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57467 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4285 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32962 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4116 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67731 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4376 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4244 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
We were soooo close to not having to do this! Oh well.

adetaylor committed:
Ref count various WebKit types within Swift
https://bugs.webkit.org/show_bug.cgi?id=302644
rdar://164888628

Reviewed by Geoffrey Garen.

This informs Swift that some C++ types should be treated as Swift reference
types (i.e. reference-counted) rather than value types. It teaches Swift
how to manipulate the reference counts for these types.

This is a temporary solution. In due course, we'll be able to inform
Swift that the base WTF::RefCounted type is the ancestor of all reference
counted types, and it won't be necessary to make these changes to
individual classes.

That's enabled by either of these changes:
rdar://131065240 - allow templated types to be ref counted
rdar://160696723 - allow method calls for ref and deref
We can't currently use this functionality in WebKit because we need
to support older compilers, and also due to a compiler
bug rdar://162195228. So for now, we will annotate just those
specific types which are required for upcoming CoreIPC message handler
work.

To avoid warnings/errors, these functions need to have their parameters
marked as _Nonnull, or specifically to adopt the WTF macro WTF_NONNULL.
That currently causes problems with -Wnullability-completeness, which
complains that other functions in the same headers are not duly
annotated. WebKit has switched off this warning for its whole codebase
but had overlooked TestWebKitAPI, which does not inherit warning flags
from the rest of the .xcconfig system. Add this warning flag there too
in order to prevent these compile errors when TestWebKitAPI indirectly
includes IPC/Connection.h

Canonical link: https://commits.webkit.org/303461@main
